modules/classloaders/hdfs-class-loader/src/main/java/org/apache/accumulo/classloader/hdfs/HDFSContextClassLoaderFactory.java

```diff
@@ -0,0 +1,402 @@
+                new IllegalStateException(String.format(
+                    "checksum: %s of downloaded jar: %s (downloaded from %s to %s) did not match "
+                        + "checksum present: %s in manifest file: %s. Consider retrying and/or "
+                        + "updating the checksum for this jar in the manifest file",
+                    computedChecksumLocalJar, jar.getJarName(), HDFSJarPath, localTempJarPath,
+                    jar.getChecksum(), HDFSManifestFile)));
+          }
+          // use final (non-temp) jar path for the URL as that is what will exist if this op
+          // completes
+          var localFinalJarPath = localFinalContextDir.resolve(jar.getJarName());
+          urls[i++] = localFinalJarPath.toUri().toURL();
+        }
+      } else {
+        throw new ContextClassLoaderException(contextName);
+      }
+    }
+
+    return new TempState(contextName, localTempContextDir, localFinalContextDir,
+        HDFSManifestFileChecksum, urls);
+  }
+
+  private Pair<String,URLClassLoader> createClassLoader(TempState tempState) throws IOException {
+    // if the final path already exists, we are replacing it with temp path, so recursively delete
+    // the final path
+    if (Files.exists(tempState.finalPath)) {
+      FileUtils.deleteDirectory(tempState.finalPath.toFile());
+    }
+    // rename temp path to final path
+    if (tempState.tempPath.toFile().renameTo(tempState.finalPath.toFile())) {
+      LOG.info("Renamed {} to {}", tempState.tempPath, tempState.finalPath);
+    }
+    var classLoader = new URLClassLoader(tempState.urls, ClassLoader.getSystemClassLoader());
+    return new Pair<>(tempState.checksum, classLoader);
+  }
+
+  public String checksumLocalFile(java.nio.file.Path path) throws IOException {
+    try (var in = FileUtils.openInputStream(path.toFile())) {
+      return checksum(in.readAllBytes());
+    }
+  }
+
+  public static String checksum(byte[] fileBytes) {
+    try {
+      StringBuilder checksum = new StringBuilder();
+      for (byte b : MessageDigest.getInstance(HASH_ALG).digest(fileBytes)) {
+        checksum.append(String.format("%02x", b));
+      }
+      return checksum.toString();
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Override
+  public ClassLoader getClassLoader(String contextName) throws ContextClassLoaderException {
+    Path HDFSContextPath = new Path(HDFSContextsDir, contextName);
+    Path HDFSManifestFile = new Path(HDFSContextPath, MANIFEST_FILE_NAME);
+
+    try (var HDFSManifestFileIn = hdfs.open(HDFSManifestFile)) {
+      // if another thread has already started creating the class loader for this context:
+      // wait a short time for it to succeed in creating and caching the classloader, otherwise
+      // continue and try to create and cache the classloader ourselves (only one result will be
+      // cached)
+      try (var allContextDirs = Files.list(localContextsDir)) {
```

Comment:
This entire try block is not using the opened manifest file. Looks like we can delay the opening of the file since we may return early or throw an exception here.

Reply:
Fixed